From f1324de380b15958860f3d638652a3390e7c0b3a Mon Sep 17 00:00:00 2001
From: Selvavignesh Shanmugam <sshanmugam@maxlinear.com>
Date: Mon, 29 Jul 2024 19:21:47 +0800
Subject: [PATCH] WLANRTSYS-83576 Skip only the free SID event to driver when
 recovery occurs

Issue:
When recovery occurs, the mxl_hostapd_free_aid is skipped, the aid is not
freed in aid_pool causing the previous AID unusable.

Fix:
Skip only the SID free driver event when recovery occurs.

FIXUP to WLANRTSYS-62845 AP MLD Auth Assoc and eapol changes
---
 src/ap/mxl_hostapd.c | 9 ++++++---
 src/ap/sta_info.c    | 1 -
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 038ada59d..4199713ba 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -310,11 +310,14 @@ void mxl_hostapd_free_aid(struct hostapd_data *hapd, struct sta_info *sta)
 		aid_pool[(sta->aid - aid_offset - 1) / 32]
 				&= ~BIT((sta->aid - aid_offset - 1) % 32);
 		/* Free SID */
-		(*hapd->driver->free_aid)(hapd->drv_priv, &sta->aid);
+		if (!mxl_hostapd_is_recovery(hapd->iface))
+			(*hapd->driver->free_aid)(hapd->drv_priv, &sta->aid);
 	} else {
 		/* Legacy */
-		sta->aid -= hapd->iconf->mbssid_aid_offset;
-		(*hapd->driver->free_aid)(hapd->drv_priv, &sta->aid);
+		if (!mxl_hostapd_is_recovery(hapd->iface)) {
+			sta->aid -= hapd->iconf->mbssid_aid_offset;
+			(*hapd->driver->free_aid)(hapd->drv_priv, &sta->aid);
+		}
 	}
 
 	sta->aid = 0;
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 6a9bdc083..89784c0e1 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -242,7 +242,6 @@ void ap_free_sta(struct hostapd_data *hapd, struct sta_info *sta)
 	ap_sta_list_del(hapd, sta);
 
 #ifdef CONFIG_VENDOR_MXL
-if (!mxl_hostapd_is_recovery(hapd->iface))
 	mxl_hostapd_free_aid(hapd, sta);
 #endif /* CONFIG_VENDOR_MXL */
 
-- 
2.43.0

